routes/web.php

```diff
@@ -17,10 +17,10 @@
 });
 
 $router->group(['middleware' => ['auth', 'version-switcher']], function () use ($router) {
-    $router->get('/list/{listId}', 'ListController@get');
-    $router->head('/list/{listId}/{materialId}', 'ListController@checkMaterial');
-    $router->put('/list/{listId}/{materialId}', 'ListController@addMaterial');
-    $router->delete('/list/{listId}/{materialId}', 'ListController@removeMaterial');
+    $router->get('/list/{type}', 'ListController@get');
```

Comment:
This is where I think this approach get's a bit iffy. We're binding to the V1 version of the spec. But what if a future version introduces a new path? Do we bind `v2/ListController` or add the route to v1 throwing NotFound?

Reply:
As far as I see it spec-wise we are introducing new spec versions as the API changes - as beforehand.
And the latest "HEAD", the symlink, point to the newest version.

Let say that in version 3 we introduce the path: GET "/list/john":
```
$router->group(['middleware' => ['auth', 'version-switcher']], function () use ($router) {
    $router->get('/list/{type}', 'ListController@get');
    $router->get('/list/john', 'ListController@john');

    $router->head('/list/{type}/{item}', 'ListController@itemAvailability');
    $router->put('/list/{type}/{item}', 'ListController@addItem');
    $router->delete('/list/{type}/{item}', 'ListController@removeItem');

    $router->put('/migrate/{openlistId}', 'MigrateController@migrate');
});
```

If the accept-version header is set to "3" the middleware will look after that method on the v3 controller and serve that.
If it is set to "2" Laravel/Lumen automatically returns a 404 Not Found if it cannot resolve the method.
And it that way the scenario you describe is automatically handled.
Or do I misunderstand you?

Reply:
Hmm, right.

The thing that rubs me the wrong way is that `ListController@get` looks like a reference to a specific method on a specific class, but it isn't really, because we make the switcheroo behind the curtain. But I can't see a nice way around it.

One thing that might work: Looks like Lumen is using the container to resolve the controller class, so the `uses` is `list_controller@get` and the VersionSwitcher sets an alias for the correct `ListController` class, it should be more obvious that it's not the specific v1 implementation we're talking about, and we're using proper DI rather than fiddling with the `uses`. Whatcha think?

Reply:
Changed the version swapping to be done in the container rather that through the route resolver.